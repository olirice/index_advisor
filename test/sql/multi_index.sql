begin;

    create extension index_advisor;

    create table author(
        id serial primary key,
        name text not null
    );

    create table publisher(
        id serial primary key,
        name text not null,
        corporate_address text
    );

    create table book(
        id serial primary key,
        author_id int not null references author(id),
        publisher_id int not null references publisher(id),
        title text
    );

    create table review(
        id serial primary key,
        book_id int references book(id),
        body text not null
    );

    select
        index_advisor('
            select
                book.id,
                book.title,
                publisher.name as publisher_name,
                author.name as author_name,
                review.body review_body
            from
                book
                join publisher
                    on book.publisher_id = publisher.id
                join author
                    on book.author_id = author.id
                join review
                    on book.id = review.book_id
            where
                author.id = $1
                and publisher.id = $2
        ');

rollback;
