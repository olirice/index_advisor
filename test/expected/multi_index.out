begin;
    create extension index_advisor version '0.1.1' cascade;
NOTICE:  installing required extension "hypopg"
    create table author(
        id serial primary key,
        name text not null
    );
    create table publisher(
        id serial primary key,
        name text not null,
        corporate_address text
    );
    create table book(
        id serial primary key,
        author_id int not null references author(id),
        publisher_id int not null references publisher(id),
        title text
    );
    create table review(
        id serial primary key,
        book_id int references book(id),
        body text not null
    );
    select
        *
    from
        index_advisor('
            select
                book.id,
                book.title,
                publisher.name as publisher_name,
                author.name as author_name,
                review.body review_body
            from
                book
                join publisher
                    on book.publisher_id = publisher.id
                join author
                    on book.author_id = author.id
                join review
                    on book.id = review.book_id
            where
                author.id = $1
                and publisher.id = $2
        ');
 startup_cost_before | startup_cost_after | total_cost_before | total_cost_after |                                                                            index_statements                                                                            
---------------------+--------------------+-------------------+------------------+------------------------------------------------------------------------------------------------------------------------------------------------------------------------
 27.26               | 12.77              | 68.48             | 42.37            | {"CREATE INDEX ON public.book USING btree (author_id)","CREATE INDEX ON public.book USING btree (publisher_id)","CREATE INDEX ON public.review USING btree (book_id)"}
(1 row)

rollback;
